### PR TITLE
Theming: Handle errors on uploaded files properly

### PR DESCRIPTION
**Pull request by juliusknorr**
Upload should fail if an error is given by PHP. A check for proper file formats is also added for the logo now.

#6581

### CODE REVIEW

**Comment by juliushaertl** [comment]
Please review @nextcloud/theming

File: apps/theming/lib/Controller/ThemingController.php [hunk 1]
```diff
@@ -207,11 +207,32 @@ public function updateLogo() {
...
+		$error = null;
+		$phpFileUploadErrors = array(
+			0 => $this->l10n->t('There is no error, the file uploaded with success'),
+			1 => $this->l10n->t('The uploaded file exceeds the upload_max_filesize directive in php.ini'),
+			2 => $this->l10n->t('The uploaded file exceeds the MAX_FILE_SIZE directive that was specified in the HTML form'),
+			3 => $this->l10n->t('The uploaded file was only partially uploaded'),
+			4 => $this->l10n->t('No file was uploaded'),
+			6 => $this->l10n->t('Missing a temporary folder'),
+			7 => $this->l10n->t('Failed to write file to disk.'),
+			8 => $this->l10n->t('A PHP extension stopped the file upload.'),
```
Comment (danxuliu) [inline apps/theming/lib/Controller/ThemingController.php]: Please replace the magic numbers with [their named constants](http://php.net/manual/en/features.file-upload.errors.php).

**Review by danxuliu — COMMENTED** [review]
Could you add tests for the upload errors (both for PHP errors and unsupported formats)? :-)

Besides that, in the UI the loading icon is not removed when an upload fails. `$('#theming_settings_loading').hide();` should be called from `fail` in both [uploadParamsLogo](https://github.com/nextcloud/server/blob/b16ce1b4babd9cd59d5797eafe31f4c96fb8149b/apps/theming/js/settings-admin.js#L141) and [uploadParamsLogin](https://github.com/nextcloud/server/blob/b16ce1b4babd9cd59d5797eafe31f4c96fb8149b/apps/theming/js/settings-admin.js#L159) (either in this pull request or in a new one, I have no preference ;-) )

**Comment by juliushaertl** [comment]
@danxuliu Thanks for the comments. Those should be addressed now. :wink:

**Comment by danxuliu** [comment]
>  Thanks for the comments. Those should be addressed now. 😉

Thanks :-)

However... may I suggest to use a data provider for `testUpdateLogoLoginScreenUploadWithInvalidImageUpload` and `testUpdateLogoUploadWithInvalidImageUpload` passing the PHP error and the expected message to cover all the possible cases? ;-)

**Comment by juliushaertl** [comment]
@danxuliu Sure. I've pushed an update. :wink:

File: apps/theming/tests/Controller/ThemingControllerTest.php [hunk 2]
```diff
@@ -452,6 +515,105 @@ public function testUpdateLogoLoginScreenUploadWithInvalidImage() {
...
+			[UPLOAD_ERR_INI_SIZE, 'The uploaded file exceeds the upload_max_filesize directive in php.ini'],
+			[UPLOAD_ERR_FORM_SIZE, 'The uploaded file exceeds the MAX_FILE_SIZE directive that was specified in the HTML form'],
+			[UPLOAD_ERR_PARTIAL, 'The uploaded file was only partially uploaded'],
+			[UPLOAD_ERR_NO_FILE, 'No file was uploaded'],
+			[UPLOAD_ERR_NO_TMP_DIR, 'Missing a temporary folder'],
+			[UPLOAD_ERR_CANT_WRITE, 'Failed to write file to disk.'],
+			[UPLOAD_ERR_EXTENSION, 'A PHP extension stopped the file upload.'],
+		];
+	}
+	/**
```
Comment (danxuliu) [inline apps/theming/tests/Controller/ThemingControllerTest.php]: Missing new line before the comment.

File: apps/theming/tests/Controller/ThemingControllerTest.php [hunk 3]
```diff
@@ -452,6 +515,105 @@ public function testUpdateLogoLoginScreenUploadWithInvalidImage() {
...
+	 * @dataProvider dataPhpUploadErrors
+	 */
+	public function testUpdateLogoLoginScreenUploadWithInvalidImageUpload($expectedError, $expectedErrorMessage) {
```
Comment (danxuliu) [inline apps/theming/tests/Controller/ThemingControllerTest.php]: I would rename `$expectedError` to `$error`. `$expectedError` would be appropriate if you were asserting that the system returned that error, but in this case the error is a precondition of the test instead (that is, you are not expecting that error, you are explicitly generating it to then check how the system behaves in that case).

File: apps/theming/tests/Controller/ThemingControllerTest.php [hunk 4]
```diff
@@ -452,6 +515,105 @@ public function testUpdateLogoLoginScreenUploadWithInvalidImage() {
...
+			],
+			Http::STATUS_UNPROCESSABLE_ENTITY
+		);
+		$this->assertEquals($expected, $this->themingController->updateLogo());
+	}
+
+	/**
+	 * @dataProvider dataPhpUploadErrors
+	 */
+	public function testUpdateLogoUploadWithInvalidImageUpload($expectedError, $expectedErrorMessage) {
```
Comment (danxuliu) [inline apps/theming/tests/Controller/ThemingControllerTest.php]: Same as above, just `$error` instead of `$expectedError`.

**Review by danxuliu — COMMENTED** [review]
>  Sure. I've pushed an update. 😉

Awesome, thanks :-)

I know that you hate me (I do too), but in that last update some minor style issues were introduced :-P It would be great if you could address them too :-) This is the last one, I promise! :-)

**Comment by juliushaertl** [comment]
@danxuliu Fixed. 😉

**Review by danxuliu — APPROVED** [review]
>  Fixed. 😉

Great, thanks! :-D

**Review by rullzer — APPROVED** [review]
Good stuff!

**Pull request merged**